### PR TITLE
Recipe for extracting number of final participants

### PR DESCRIPTION
**Pull request by CG80499**
- Added FinalN recipe
- Added monoT5 agent for paragraph classification

### CODE REVIEW

File: ice/recipes/finalN.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+        .replace("..", "")
+    )
+    number_strings = list(map(remove_parentheses, number_strings))
+
+    # Remove "," or "." from the end of the number string
+    for i, number in enumerate(number_strings):
+        if number[-1] == "," or number[-1] == ".":
+            number_strings[i] = number[:-1]
+
+    return number_strings
```
Comment (alexmojaki) [inline ice/recipes/finalN.py:150]:
```suggestion
    numbers_set = set("0123456789")
    return [
        word.replace("(", "")
            .replace(")", "")
            .replace("...", "")
            .replace("..", "")
            .rstrip(",.")
        for word in words
        if numbers_set.intersection(word)
    ]
```

1. This assumes that removing more than one trailing `,` or `.` is OK, which your original code does not.
2. Regexes or an actual library might be better suited here.
3. Is it fine to return strings containing numbers that clearly aren't numbers, like `"agley-2021.pdf"`?

File: ice/recipes/finalN.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        final_answer=answer,
+        full_response=response_text.strip(),
+    )
+    return answer, token_usage
+
+
+@trace
+async def get_best_answer(abstract: str, text: str, population: str):
+    prompt = answer_populations_question_with_reasoning(abstract, text, population)
+    answers = [(await answer_with_reasoning(prompt, cache_id=i)) for i in range(5)]
```
Comment (alexmojaki) [inline ice/recipes/finalN.py:234]: Should these happen in parallel?

File: ice/recipes/finalN.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+    total_token_usage = sum([t for _, t in answers])
+    log.warn(f"Total token usage: {total_token_usage}")
```
Comment (alexmojaki) [inline ice/recipes/finalN.py:236]: `warn` sounds a bit heavy for something expected?

File: ice/recipes/finalN.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+            score += 0.01 * len(numbers)
+            scores.append(score)
+        else:
+            scores.append(0)
+
+    answers = [(answer, score) for answer, score in zip(answers, scores)]
+    answers.sort(key=lambda x: x[1], reverse=True)
+
+    answers = [a for a, _ in answers]
+    return answers[0]
```
Comment (alexmojaki) [inline ice/recipes/finalN.py:265]:
```suggestion
    answer_counts_by_number = collections.Counter(
        number
        for answer_numbers in numbers_in_answers
        for number in set(answer_numbers)  # only count each number once per answer
    )

    def score(answer_numbers):
        if answer_numbers:
            average_count = statistics.mean(
                answer_counts_by_number[number] for number in answer_numbers
            )
        else:
            average_count = 0
        return average_count, len(answer_numbers)

    answer, _ = max(
        zip(answers, numbers_in_answers),
        key=lambda t: score(t[1]),
    )
    return answer
```

**Review by alexmojaki — COMMENTED** [review]
Hello :wave: Please don't take this 'review' too seriously. I'm not an Ought employee, I'm just passing by.

**Review by lslunis — CHANGES_REQUESTED** [review]
I'd like to get ice/nn/weights out of this repo so that people can get started with a smaller download. That means:

1. We should do something like download them from S3 instead. I'm going to set that up soon, but might not get to it this week, and this PR shouldn't go in until that's done
2. I'm going to **rewrite git history** to make the repo smaller (and then obviously never do that again, because it's terrible). That might cause problems with eventually merging this PR, and I can help with that when the time comes

**Comment by alexmojaki** [comment]
> I'd like to get ice/nn/weights out of this repo so that people can get started with a smaller download.

I'm not saying you shouldn't do this, but if you're concerned about this when the model weights are 238KB, then I think I should point out that I ran out of disk space because the Python dependencies take up 4.5 GB, of which 3.3 GB is `torch`.

**Comment by alexmojaki** [comment]
Ah, I see you're probably talking more about the existing 6.4 MB than what's in this PR, that changes things.

**Comment by lslunis** [comment]
Looks like when I force-pushed main, it caused this PR to close(!). I'm going to reopen it tomorrow or Sunday based on the new main.

**Pull request closed without merging**